Window to worker communication series (#4340)
GoogleChrome/web.dev: src/site/content/en/reliable/two-way-communication-guide/index.md

tags:
  - service-worker
  - offline
---

In some cases, a web app might need to establish a **two-way** communication channel between the
page and the service worker.

For example: in a podcast PWA one could build a feature to let the user [download episodes for
offline consumption](https://web.dev/app-like-pwas/#proactive-background-downloading) and allow the
service worker to keep the page regularly informed about the progress, so the [main
thread](https://developer.mozilla.org/en-US/docs/Glossary/Main_thread) can update the UI.

In this guide we'll explore the different ways of implementing a **two-way** communication between
the [Window](https://developer.mozilla.org/en-US/docs/Web/API/Window) and [service
worker](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) context, by exploring
different APIs, the [Workbox library](https://developers.google.com/web/tools/workbox), as well as
some advanced cases.

<figure class="w-figure">
  <img src="two-way-communication-diagram.png"
       width="800"
       alt="Diagram showing a service worker and the page exchanging messages.">

> </figure>
> 
> ## Using Workbox {: #using-workbox }

Reviewer: ```suggestion
</figure>

{% Aside %}
  Check out [Workers overview](/workers-overview/) for a high-level explanation
  of when to use web workers versus service workers and the rest of the
  [Window to workers communication](/reliable/#window-to-worker-communication) series
  for guides on other common use cases.
{% endAside %}

## Using Workbox {: #using-workbox }
```

---
layout: post 

> title: "Two way communication" 

Reviewer: ```suggestion
title: "Two-way communication with service workers" 
```

tags:
  - blog
  - service-worker
  - web-worker
---

In some cases, a web app might need to establish a **two-way** communication channel between the
page and the service worker.

For example: in a podcast PWA one could build a feature to let the user [download episodes for
offline consumption](https://web.dev/app-like-pwas/#proactive-background-downloading) and allow the
service worker to keep the page regularly informed about the progress, so the [main
thread](https://developer.mozilla.org/en-US/docs/Glossary/Main_thread) can update the UI.

In this guide we'll explore the different ways of implementing a **two-way** communication between
the [Window](https://developer.mozilla.org/en-US/docs/Web/API/Window) and [service
worker](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) context, by exploring
different APIs, the [Workbox library](https://developers.google.com/web/tools/workbox), as well as
some advanced cases.

<figure class="w-figure">
  <img src="two-way-communication-diagram.png"
       width="800"
       alt="Diagram showing a service worker and the page exchanging messages.">

> </figure>

Reviewer: Is this diagram complete? 

![Uploading image.png…]()

Author: It's the diagram that Andrew used in this part of the the talk, so, yes. Maybe it doesn't have text as in others. Do you think it's necessary?

## Using Workbox

[Workbox-window](https://developers.google.com/web/tools/workbox/modules/workbox-window) is a set of
modules of the [Workbox library](https://developers.google.com/web/tools/workbox) that are intended
to run in the window context. The [Workbox
class](https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-window.Workbox)

> provides a `messageSW()` method, to `postMessage()` the instance's registered service worker and
> await a response.

Reviewer: ```suggestion
provides a `messageSW()` method to send a message to the instance's registered service worker and
await a response.
```

> The following page code, creates a new `Workbox` instance and sends a message to the service worker

Reviewer: ```suggestion
The following page code creates a new `Workbox` instance and sends a message to the service worker
```

> to obtain its version. The result of the operation is logged via console:

Reviewer: ```suggestion
to obtain its version:
```
I think we can assume our audience is familiar with logging

```javascript
const wb = new Workbox('/sw.js'); 
wb.register(); 

const swVersion = await wb.messageSW({type: 'GET_VERSION'}); 
console.log('Service Worker version:', swVersion);
```

The service worker implements a message listener on the other end, and responds to the registered
service worker:

```javascript
const SW_VERSION = '1.0.0';

addEventListener('message', (event) => {
  if (event.data.type === 'GET_VERSION') {
    event.ports[0].postMessage(SW_VERSION);
  }
});
```

Under the hood the library uses a browser API that we'll review in the next section: [Message
Channel](https://developer.mozilla.org/en-US/docs/Web/API/Channel_Messaging_API), but abstracts many
implementation details, making it easier to use, while leveraging the [wide browser

> support](https://caniuse.com/?search=messagechannel) this API counts.

Reviewer: ```suggestion
support](https://caniuse.com/mdn-api_messagechannel_port1) this API has.
```
`?search` URLs on Can I Use are unstable, click the hashtag to get a stable link

<figure class="w-figure">
  <img src="workbox-window-diagram.png"
       width="700"
       alt="Diagram showing two-way communication between page and service worker, using Workbox Window.">
</figure>

## Using Browser APIs

If the Workbox library is not enough for your needs, there are several lower-level APIs available to
implement **“two-way”** communication between pages and service workers. They have some similarities
and differences:

Similarities:

- In all cases the communication starts on one end via the `postMessage()` interface and is received

>   on the other end by implementing an `onmessage` handler.

Reviewer: `message` not `onmessage`

Author: Done

- In practice, all the available APIs allow us to implement the same use cases, but some of them
  might simplify development in some scenarios.

Differences:

- They have different ways of identifying the other side of the communication: some of them use an
  explicit reference to the other context, while others can communicate implicitly via a proxy
  object instantiated on each side.

> - Browser support varies among them, so check what is the level of support of a specific API before
>   using it.

Reviewer: ```suggestion
- Browser support varies among them.
```

<figure class="w-figure">
  <img src="communication-apis.png"
       width="600"
       alt="Diagram showing two-way communication between page and service worker, and the available browser APIs.">
</figure>

### Broadcast Channel API

The [Broadcast Channel API](https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API)
allows basic communication between browsing contexts via [BroadcastChannel
objects](https://developer.mozilla.org/en-US/docs/Web/API/BroadcastChannel). 

To implement it, first, each context has to instantiate a `BroadcastChannel` object with the same ID
and send and receive messages from it:

```javascript
const broadcast = new BroadcastChannel('channel-123'); 
```

> The BroadcastChannel object exposes a `postMessage()` interface to send a message any listening

Reviewer: ```suggestion
The BroadcastChannel object exposes a `postMessage()` interface to send a message to any listening
```

  if (event.data && event.data.type === 'MSG_ID') {
    //process response
  }
};
```

Similarly, the service worker listens to messages by implementing an `onmessage` listener:

```javascript
//listen to messages
self.addEventListener('message', (event) => {
  if (event.data && event.data.type === 'MSG_ID') {
    //Process message
  }
});
```

To communicate back with any of its clients, the service worker obtains an array of
[WindowClient](https://developer.mozilla.org/en-US/docs/Web/API/WindowClient) objects by executing
methods such as
[Clients.matchAll()](https://developer.mozilla.org/en-US/docs/Web/API/Clients/matchAll) and
[Clients.get()](https://developer.mozilla.org/en-US/docs/Web/API/Clients/get). Then it can
`postMessage()` any of them:

```javascript
//Obtain an array of Window client objects
self.clients.matchAll(options).then(function (clients) {
  if (clients && clients.length) {
    //Respond to last focused tab
    clients[0].postMessage({type: 'MSG_ID'});
  }
});
```

<figure class="w-figure">
  <img src="client-api.png"
       width="500"
       alt="Diagram showing a service worker communicating with an array of clients.">
</figure>

> `Client API` is a good option to communicate easily with all the active tabs from a service worker

Reviewer: Formatting of Client API isn't consistent with earlier

Author: I see in all cases it appears as `Client API`. I can't find the mismatch here.

HTTP 203](https://www.youtube.com/watch?v=9UNwHmagedE&feature=youtu.be&t=697) to know more about it.
{% endAside %}

### Message Channel

[Message Channel](https://developer.mozilla.org/en-US/docs/Web/API/Channel_Messaging_API) requires
defining and passing a port from one context to another to establish a **two-way** communication
channel.

To initialize the channel, the page instantiates a [MessageChannel
object](https://developer.mozilla.org/en-US/docs/Web/API/MessageChannel/MessageChannel) and uses it
to send a port to the registered service worker. The page also implements an `onmessage` listener on
it to receive messages from the other context:

```javascript
const messageChannel = new MessageChannel();

//Init port
navigator.serviceWorker.controller.postMessage({type: 'PORT_INITIALIZATION'}, [
  messageChannel.port2,
]);

//Listen to messages
messageChannel.port1.onmessage = (event) => {
  // Process message
};
```

<figure class="w-figure">
  <img src="message-channel.png"
       width="600"
       alt="Diagram showing a page passing a port to a service worker, to establish two-way communication.">
</figure>

> The service worker receives the port, saves a reference to it and uses it to postMessage the other

Reviewer: ```suggestion
The service worker receives the port, saves a reference to it and uses it to send a message to the other
```

side:

```javascript
let communicationPort;

//Save reference to port
self.addEventListener('message', (event) => {
  if (event.data && event.data.type === 'PORT_INITIALIZATION') {
    communicationPort = event.ports[0];
  }
});

//Send messages
communicationPort.postMessage({type: 'MSG_ID'});
```

`MessageChannel` is currently supported by [all major
browsers](https://caniuse.com/?search=channel).

### Advanced APIs: Background Sync and Background Fetch

In this guide we explored ways of implementing **two-way** communication techniques, for relatively
simple cases, like passing a string message describing the operation to perform, or a list of URLs
to cache from one context to the other. In this section we'll explore two APIs to handle specific
scenarios: lack of connectivity and long downloads.

#### Background Sync

A chat app might want to make sure that messages are never lost due to bad connectivity. The
[Background Sync API](https://developers.google.com/web/updates/2015/12/background-sync) lets you
defer actions to be retried when the user has stable connectivity. This is useful for ensuring that
whatever the user wants to send, is actually sent. 

> Instead of the `postMessage()` interface, the page register for a `sync`:

Reviewer: ```suggestion
Instead of the `postMessage()` interface, the page registers a `sync`:
```

```javascript
navigator.serviceWorker.ready.then(function (swRegistration) {
  return swRegistration.sync.register('myFirstSync');
});
```

The service worker then listens for the `sync` event to process the message:

```javascript
self.addEventListener('sync', function (event) {
  if (event.tag == 'myFirstSync') {
    event.waitUntil(doSomeStuff());
  }
});
```

The function `doSomeStuff()` should return a promise indicating the success/failure of whatever it's
trying to do. If it fulfills, the sync is complete. If it fails, another sync will be scheduled to
retry. Retry syncs also wait for connectivity, and employ an exponential back-off. 

Once the operation has been performed, the service worker can then communicate back with the page to
update the UI, by using any of the communication APIs explored earlier.

> Google search uses `Background Sync` to persist failed queries due to bad connectivity, and retry

Reviewer: Formatting of Background Sync is not consistent with earlier

Author: I see in all cases it appears as `Background Sync`. I can't find the mismatch here.

Reviewer: Oh, sorry. It's the fact that you formatted it as inline code:

    `Background Sync`

Whereas everywhere else you did not use inline code:

    Background Sync

Reviewer: ```suggestion
Google search uses Background Sync to persist failed queries due to bad connectivity, and retry
```

them later when the user is online. Once the operation is performed, they communicate the result to
the user via a web push notification:

<figure class="w-figure">
  <img src="google-search-background-sync.png"
       width="700"
       alt="Diagram showing a page passing a port to a service worker, to establish two-way communication.">
</figure>

{% Aside %} Check out [this guide](https://web.dev/resilient-search-experiences/) to see this

> feature in your site using [Workbox Background

Reviewer: "Check out Resilient search experiences to learn how to implement this feature…"?

Author: Done

Sync](https://developers.google.com/web/tools/workbox/modules/workbox-background-sync). {% endAside
%}

#### Background Fetch

For relatively short bits of work like sending a message, or a list of URLs to cache, the options
explored so far are a good choice. If the task takes too long the browser will kill the service
worker, otherwise it's a risk to the user's privacy and battery. 

The [Background Fetch API](https://developers.google.com/web/updates/2018/12/background-fetch)

> allows you to offload a long task to a service worker, like downloading a movie, podcasts, or levels

Reviewer: ```suggestion
allows you to offload a long task to a service worker, like downloading movies, podcasts, or levels
```
https://owl.purdue.edu/owl/general_writing/mechanics/parallel_structure.html#:~:text=Parallel%20structure%20means%20using%20the,and%22%20or%20%22or.%22

of a game. 

> To communicate the page with the service worker, use `backgroundFetch.fetch`, instead of

Reviewer: ```suggestion
To communicate to the service worker from the page, use `backgroundFetch.fetch`, instead of
```

    'my-fetch',
    ['/ep-5.mp3', 'ep-5-artwork.jpg'],
    {
      title: 'Episode 5: Interesting things.',
      icons: [
        {
          sizes: '300x300',
          src: '/ep-5-icon.png',
          type: 'image/png',
        },
      ],
      downloadTotal: 60 * 1024 * 1024,
    },
  );
});
```

The `BackgroundFetchRegistration` object allows the page listen to the `progress` event to follow
the progress of the download:

```javascript
bgFetch.addEventListener('progress', () => {
  // If we didn't provide a total, we can't provide a %.
  if (!bgFetch.downloadTotal) return;

  const percent = Math.round(
    (bgFetch.downloaded / bgFetch.downloadTotal) * 100,
  );
  console.log(`Download progress: ${percent}%`);
});
```

<figure class="w-figure">
  <img src="podcast-pwa.png"
       width="700"
       alt="Diagram showing a page passing a port to a service worker, to establish two-way communication.">
    <figcaption class="w-figcaption">The UI is updated to indicate the progress of a download (left). Thanks to service workers, the operation can continue running when all tabs have been closed (right).
    </figcaption>
</figure>

> {% Aside %} Check the [Background Fetch

Reviewer: ```suggestion
{% Aside %} Check out the [Background Fetch
```

page and the service worker.

For example: in a podcast PWA one could build a feature to let the user [download episodes for
offline consumption](https://web.dev/app-like-pwas/#proactive-background-downloading) and allow the
service worker to keep the page regularly informed about the progress, so the [main
thread](https://developer.mozilla.org/en-US/docs/Glossary/Main_thread) can update the UI.

In this guide we'll explore the different ways of implementing a **two-way** communication between
the [Window](https://developer.mozilla.org/en-US/docs/Web/API/Window) and [service
worker](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) context, by exploring
different APIs, the [Workbox library](https://developers.google.com/web/tools/workbox), as well as
some advanced cases.

<figure class="w-figure">
  <img src="two-way-communication-diagram.png"
       width="800"
       alt="Diagram showing a service worker and the page exchanging messages.">
</figure>

{% Aside %}
  Check out [Workers overview](/workers-overview/) for a high-level explanation
  of when to use web workers versus service workers and the rest of the

>   [Window to workers communication](/reliable/#window-to-worker-communication) series

Reviewer: ```suggestion
  [Communicate with workers](/reliable/#communicate-with-workers) series
```
